ksqldb-engine/src/main/java/io/confluent/ksql/utilization/PersistentQuerySaturationMetrics.java

```diff
@@ -0,0 +1,334 @@
+              ImmutableMap.of(THREAD_ID, name)
+          )
+      ));
+    }
+
+    private void reportQuerySaturation(
+        final Instant now,
+        final double saturation,
+        final QueryId queryId,
+        final MetricsReporter reporter
+    ) {
+      LOGGER.info("Reporting query saturation {} for {}", saturation, queryId);
+      reporter.report(ImmutableList.of(
+          new DataPoint(
+              now,
+              QUERY_SATURATION,
+              saturation,
+              ImmutableMap.of(QUERY_ID, queryId.toString())
+          )
+      ));
+    }
+
+    private Map<String, Map<String, Metric>> metricsByThread(final KafkaStreams kafkaStreams) {
+      return kafkaStreams.metrics().values().stream()
+          .filter(m -> m.metricName().group().equals(STREAMS_THREAD_METRICS_GROUP))
+          .collect(Collectors.groupingBy(
+              m -> m.metricName().tags().get(THREAD_ID),
+              Collectors.toMap(m -> m.metricName().name(), m -> m, (a, b) -> a)
+          ));
+    }
+
+    private Optional<Double> measure(
+        final Instant now,
+        final KafkaStreams kafkaStreams,
+        final MetricsReporter reporter
+    ) {
+      final Map<String, Map<String, Metric>> byThread = metricsByThread(kafkaStreams);
+      Optional<Double> saturation = Optional.of(0.0);
+      for (final Map.Entry<String, Map<String, Metric>> entry : byThread.entrySet()) {
+        final String threadName = entry.getKey();
+        final Map<String, Metric> metricsForThread = entry.getValue();
+        if (!metricsForThread.containsKey(STREAMS_TOTAL_BLOCKED_TIME)
+            || !metricsForThread.containsKey(STREAMS_THREAD_START_TIME)) {
+          LOGGER.info("Missing required metrics for thread: {}", threadName);
+          continue;
+        }
+        final double totalBlocked =
+            (double) metricsForThread.get(STREAMS_TOTAL_BLOCKED_TIME).metricValue();
+        final long startTime =
+            (long) metricsForThread.get(STREAMS_THREAD_START_TIME).metricValue();
+        final ThreadSaturation threadSaturation = perThreadSaturation.computeIfAbsent(
+            threadName,
+            k -> {
+              LOGGER.debug("Adding saturation for new thread: {}", k);
+              return new ThreadSaturation(threadName, startTime, window, sampleMargin);
+            }
+        );
+        LOGGER.debug("Record thread {} sample {} {}", threadName, totalBlocked, startTime);
+        final BlockedTimeSample blockedTimeSample = new BlockedTimeSample(now, totalBlocked);
+        final Optional<Double> measured = threadSaturation.measure(now, blockedTimeSample);
+        LOGGER.debug(
+            "Measured value for thread {}: {}",
+            threadName,
+            (measured.map(Object::toString).orElse(""))
+        );
+        measured.ifPresent(m -> reportThreadSaturation(now, m, threadName, reporter));
+        saturation = compareSaturation(saturation, measured) > 0 ? saturation : measured;
```

Comment:
To see if I'm following - this basically picks the larger of saturation or measured, right? So as we go through the loop to calculate thread metrics, we'll track the largest thread saturation and then ultimately report that when we exit the for loop. How would saturation ever not exist? We set it to `Optional.of(0.0)` to begin with so if the first measured doesn't exist we'd return -1 and reset saturation to `Optional.of(0.0)`, right? So is the check for if the first parameter exits just for when we compare saturation in `run()`?

Reply:
It may not exist if we don't have good data to compute saturation. This agent runs every minute and samples the total blocked time for each stream thread. It then computes the blocked time for the current window by looking back five minutes and taking the difference between the current value and that sample. However, it's possible that there isn't a good (as defined as a sample that's within sampleMargin) sample from five minutes ago (e.g. if the thread couldn't run for some reason). If that happens, we don't know the current saturation value so we return Optional.empty().